Add block restriction to phase field actions that define variables

GiudGiud opened the pull request:
Responding to https://github.com/idaholab/moose/discussions/23593

Closes #23632

GiudGiud committed:
Make KKS action error since it s not implemented

GiudGiud committed:
Add block restriction to all phase field actions that define variables
All the ones that define kernels will just inherit have them use the block restriction from the variable they use, reducing potential for error
refs #23632

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1392520/) on 1dfa074 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23633/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1392534/) on 1dfa074 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>cef69d</th>
      <th colspan=4>#23633 1dfa07</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.78%</td>
      <td>84.79%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>83653</td>
      <td>83657</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15015</td>
      <td>15011</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23633/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23633/coverage/framework)


## Modules coverage

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>cef69d</th>
      <th colspan=4>#23633 1dfa07</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.58%</td>
      <td>89.46%</td>
      <td>-0.12%</td>
      <td>47.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12460</td>
      <td>12477</td>
      <td>+17</td>
      <td>19</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1450</td>
      <td>1470</td>
      <td>+20</td>
      <td>21</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23633/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23633/coverage/phase_field)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23633/coverage/xfem)

</details>

## Warnings

- `phase_field` new line coverage rate 47.50% is less than the suggested 90.0%



This comment will be updated on new commits.

dschwen commented on modules/phase_field/src/actions/NonconservedAction.C:
```diff
@@ -97,7 +101,6 @@ NonconservedAction::act()
     InputParameters params1 = _factory.getValidParams(kernel_type);
     params1.set<NonlinearVariableName>("variable") = _var_name;
     params1.applyParameters(parameters());
-
```
I'd prefer it if unnecessary white space changes we reduced to a minimum

GiudGiud replied:
removed them from the first commit

dschwen commented on modules/phase_field/src/actions/PolycrystalVariablesAction.C:
```diff
@@ -75,6 +77,9 @@ PolycrystalVariablesAction::act()
 
       var_params.applySpecificParameters(_pars, {"order", "family"});
       var_params.set<std::vector<Real>>("scaling") = {_pars.get<Real>("scaling")};
+      if (isParamValid("block"))
+        var_params.set<std::vector<SubdomainName>>("block") =
+            getParam<std::vector<SubdomainName>>("block");
```
Would this not work?

```suggestion
      var_params.applySpecificParameters(_pars, {"order", "family", "block"});
      var_params.set<std::vector<Real>>("scaling") = {_pars.get<Real>("scaling")};
```

dschwen commented on modules/phase_field/src/actions/PFCRFFVariablesAction.C:
```diff
@@ -61,6 +62,9 @@ PFCRFFVariablesAction::act()
 
   var_params.applySpecificParameters(_pars, {"family", "order"});
   var_params.set<std::vector<Real>>("scaling") = {getParam<Real>("scaling")};
+  if (isParamValid("block"))
+    var_params.set<std::vector<SubdomainName>>("block") =
+        getParam<std::vector<SubdomainName>>("block");
```
```suggestion
  var_params.applySpecificParameters(_pars, {"family", "order", "block"});
  var_params.set<std::vector<Real>>("scaling") = {getParam<Real>("scaling")};
```

? and so on

GiudGiud replied:
never seen that. I ll try

GiudGiud committed:
Address Daniel's review:
Use apply specific parameter instead of the isParamValid then getParam pattern
refs #23632

GiudGiud commented on modules/phase_field/src/actions/BicrystalBoundingBoxICAction.C:
```diff
@@ -72,9 +72,6 @@ BicrystalBoundingBoxICAction::act()
       poly_params.set<Real>("inside") = 0.0;
       poly_params.set<Real>("outside") = 1.0;
     }
-    if (isParamValid("block"))
```
if my understanding is correct (and I just checked also) the applyParameters above does the job of adding block

GiudGiud commented on modules/phase_field/src/actions/PolycrystalVoronoiVoidICAction.C:
```diff
@@ -57,9 +57,6 @@ PolycrystalVoronoiVoidICAction::act()
         getParam<UserObjectName>("polycrystal_ic_uo");
-    if (isParamValid("block"))
-      poly_params.set<std::vector<SubdomainName>>("block") =
-          getParam<std::vector<SubdomainName>>("block");
```
same here

GiudGiud reviewed (commented):
@dschwen your review should be addressed

dschwen commented on modules/phase_field/src/actions/DisplacementGradientsAction.C:
```diff
@@ -66,7 +69,9 @@ DisplacementGradientsAction::act()
     InputParameters params = _factory.getValidParams("StrainGradDispDerivatives");
     params.set<std::vector<VariableName>>("displacement_gradients") = _displacement_gradients;
-    params.set<std::vector<SubdomainName>>("block") = {"0"}; // TODO: add parameter for this
+    if (isParamValid("block"))
+      params.set<std::vector<SubdomainName>>("block") =
+          getParam<std::vector<SubdomainName>>("block");
```
```suggestion
    params.applySpecificParameters(parameters(), {"block"});
```

dschwen commented on modules/phase_field/src/actions/PolycrystalRandomICAction.C:
```diff
@@ -53,12 +53,9 @@ PolycrystalRandomICAction::act()
     poly_params.set<VariableName>("variable") = _var_name_base + Moose::stringify(op);
-    poly_params.set<unsigned int>("op_num") = _op_num;
     poly_params.set<unsigned int>("op_index") = op;
     poly_params.set<unsigned int>("random_type") = _random_type;
```
quite annoying that the parameter type is unsigned int here, but that's a separate issue

dschwen commented on modules/phase_field/src/actions/Tricrystal2CircleGrainsICAction.C:
```diff
@@ -67,9 +67,7 @@ Tricrystal2CircleGrainsICAction::act()
-    if (isParamValid("block"))
-      poly_params.set<std::vector<SubdomainName>>("block") =
-          getParam<std::vector<SubdomainName>>("block");
+    poly_params.applySpecificParameters(parameters(), {"block"});
```
```suggestion
    poly_params.applySpecificParameters(parameters(), {"op_num", "block"});
```

dschwen commented on modules/phase_field/src/actions/Tricrystal2CircleGrainsICAction.C:
```diff
@@ -67,9 +67,7 @@ Tricrystal2CircleGrainsICAction::act()
     poly_params.set<VariableName>("variable") = var_name;
     poly_params.set<unsigned int>("op_num") = _op_num;
```
```suggestion
```

GiudGiud committed:
Apply suggestions from code review

More applying parameters

Co-authored-by: Daniel Schwen <daniel@schwen.de>
